File: udev/udev_monitor.cpp
// Copyright (c) 2010, 2011 Sebastian Wiesner <lunaryorn@googlemail.com>

// This program is free software. It comes without any warranty, to
// the extent permitted by applicable law. You can redistribute it
// and/or modify it under the terms of the Do What The Fuck You Want
// To Public License, Version 2, as published by Sam Hocevar. See
// http://sam.zoy.org/wtfpl/COPYING for more details.

/**
 * @brief Monitor devices with libudev and Qt.
 */

#include "udev_monitor.h"
#include "qudevobserver.h"

#include <QtCore/QTextStream>
#include <QtCore/QByteArray>
#include <QtCore/QCoreApplication>

void MouseMonitor::mouseChanged(const udev_device_ptr &device) {
    QTextStream out(stdout);
    QByteArray action = udev_device_get_action(device);
    QByteArray isMouse = udev_device_get_property_value(
        device, "ID_INPUT_MOUSE");
    QByteArray sysName = udev_device_get_sysname(device);
    if (isMouse == "1" && sysName.indexOf("event") == 0) {
        udev_device_ptr parent = udev_device_get_parent(device);
        // device pointer has no initial reference, so forcibly acquire
        // a reference
        parent.ref();
        if (parent) {
            QByteArray name = udev_device_get_property_value(
                parent, "NAME");
            if (!name.isEmpty()) {
                if (action == "add") {
                    out << name << " added" << endl;
                } else if (action == "remove") {
                    out << name << " removed" << endl;
                }
            }
        }
    }
}

int main(int argc, char *argv[]) {
    QCoreApplication app(argc, argv);
    udev_ptr context = udev_new();
    udev_monitor_ptr monitor = udev_monitor_new_from_netlink(context, "udev");
    udev_monitor_filter_add_match_subsystem_devtype(monitor, "input", 0);
    QUDevObserver observer(monitor);
    MouseMonitor client;
    QObject::connect(&observer, SIGNAL(deviceChanged(const udev_device_ptr&)),
                     &client, SLOT(mouseChanged(const udev_device_ptr&)));
    return app.exec();
}

#include "udev_monitor.moc"

File: udev/udev_monitor.py
#!/usr/bin/env python2
# Copyright (c) 2011 Sebastian Wiesner <lunaryorn@googlemail.com>

# Permission is hereby granted, free of charge, to any person obtaining a
# copy of this software and associated documentation files (the "Software"),
# to deal in the Software without restriction, including without limitation
# the rights to use, copy, modify, merge, publish, distribute, sublicense,
# and/or sell copies of the Software, and to permit persons to whom the
# Software is furnished to do so, subject to the following conditions:

# The above copyright notice and this permission notice shall be included in
# all copies or substantial portions of the Software.

# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
# IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
# FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
# THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
# LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
# FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
# DEALINGS IN THE SOFTWARE.

"""
    Monitor mouse devices with udev.

    .. moduleauthor::  Sebastian Wiesner  <lunaryorn@googlemail.com>
"""

from __future__ import (print_function, division, unicode_literals,
                        absolute_import)

import sys
from signal import signal, SIGINT
from functools import partial

from pyudev import Context, Monitor
from pyudev.pyside import QUDevMonitorObserver
from PySide.QtCore import QCoreApplication


def print_mouse_status(device, status_message):
    try:
        is_mouse = device.asbool('ID_INPUT_MOUSE')
    except KeyError:
        pass
    else:
        if is_mouse and device.sys_name.startswith('event'):
            name = device.parent['NAME']
            print(name, status_message)


def main():
    app = QCoreApplication(sys.argv)
    context = Context()
    monitor = Monitor.from_netlink(context)
    monitor.filter_by(subsystem='input')
    observer = QUDevMonitorObserver(monitor)
    observer.deviceAdded.connect(
        partial(print_mouse_status, status_message="added"))
    observer.deviceRemoved.connect(
        partial(print_mouse_status, status_message="removed"))
    monitor.start()

    app.exec_()


if __name__ == '__main__':
    main()

